File: hamming.py
def calculate_parity_bits(m):
    # Calculate the number of parity bits needed
    r = 0
    while (2**r) < (m + r + 1):
        r += 1
    return r

def insert_parity_bits(data, r):
    n = len(data) + r
    hamming_code = ['P' if (i & (i + 1)) == 0 else 'D' for i in range(n)]

    # Fill data bits
    j = 0
    for i in range(n):
        if hamming_code[i] == 'D':
            hamming_code[i] = data[j]
            j += 1

    return hamming_code

def set_parity_bits(hamming_code):
    n = len(hamming_code)
    for i in range(n):
        if hamming_code[i] == 'P':
            parity_pos = i + 1
            count = 0
            for j in range(parity_pos - 1, n, 2 * parity_pos):
                count += sum(1 for bit in hamming_code[j:j + parity_pos] if bit == '1')
            hamming_code[i] = '0' if count % 2 == 0 else '1'
    return hamming_code

def generate_hamming_code(data):
    r = calculate_parity_bits(len(data))
    hamming_code = insert_parity_bits(data, r)
    hamming_code = set_parity_bits(hamming_code)
    return ''.join(hamming_code)

# Example: Data to encode (8 bits)
data = '10110011'
hamming_code = generate_hamming_code(data)
print("Generated Hamming Code for 8-bit input:", hamming_code)

File: hamming.c
#include <stdio.h>
#include <string.h>
#include <math.h>

int calculate_parity_bits(int m) {
    int r = 0;
    while ((1 << r) < (m + r + 1)) {
        r++;
    }
    return r;
}

void insert_parity_bits(char *data, int r, char *hamming_code) {
    int n = strlen(data) + r;
    int data_index = 0;

    for (int i = 0; i < n; i++) {
        if ((i & (i + 1)) == 0) {
            hamming_code[i] = 'P'; // Placeholder for parity bits
        } else {
            hamming_code[i] = data[data_index++];
        }
    }
}

void set_parity_bits(char *hamming_code, int n) {
    for (int i = 0; i < n; i++) {
        if (hamming_code[i] == 'P') {
            int parity_pos = i + 1;
            int count = 0;
            for (int j = parity_pos - 1; j < n; j += 2 * parity_pos) {
                for (int k = j; k < j + parity_pos && k < n; k++) {
                    if (hamming_code[k] == '1') {
                        count++;
                    }
                }
            }
            hamming_code[i] = (count % 2 == 0) ? '0' : '1';
        }
    }
}

void generate_hamming_code(char *data) {
    int r = calculate_parity_bits(strlen(data));
    int n = strlen(data) + r;
    char hamming_code[n + 1];
    hamming_code[n] = '\0';

    insert_parity_bits(data, r, hamming_code);
    set_parity_bits(hamming_code, n);

    printf("Generated Hamming Code: %s\n", hamming_code);
}

int main() {
    char data[] = "10110011"; // Example 8-bit data
    generate_hamming_code(data);
    return 0;
}
